src/celengine/fisheyeprojectionmode.cpp

```diff
@@ -0,0 +1,100 @@
+// fisheyeprojectionmode.cpp
+//
+// Copyright (C) 2023-present, Celestia Development Team.
+//
+// This program is free software; you can redistribute it and/or
+// modify it under the terms of the GNU General Public License
+// as published by the Free Software Foundation; either version 2
+// of the License, or (at your option) any later version.
+
+#include "fisheyeprojectionmode.h"
+#include <celmath/geomutil.h>
+
+using namespace Eigen;
+using namespace celmath;
+
+namespace celengine
+{
+
+FisheyeProjectionMode::FisheyeProjectionMode(float width, float height) : ProjectionMode(), width(width), height(height)
+{
+}
+
+Matrix4f FisheyeProjectionMode::getProjectionMatrix(float nearZ, float farZ, float zoom) const
+{
+    float aspectRatio = width / height;
+    return Ortho(-aspectRatio, aspectRatio, -1.0f, 1.0f, nearZ, farZ);
+}
+
+float FisheyeProjectionMode::getMinimumFOV() const
+{
+    return degToRad(179.99f);
+}
+
+float FisheyeProjectionMode::getMaximumFOV() const
+{
+    return degToRad(179.99f);
+}
+
+float FisheyeProjectionMode::getFOV(float zoom) const
+{
+    return degToRad(179.99f);
+}
+
+float FisheyeProjectionMode::getZoom(float fov) const
+{
+    return 1.0f;
+}
+
+float FisheyeProjectionMode::getPixelSize(float zoom) const
+{
+    return 2.0f / height;
+}
+
+double FisheyeProjectionMode::getViewConeAngleMax(float zoom) const
+{
+    return std::cos(static_cast<double>(getFOV(zoom)) / 2.0);
+}
+
+Vector3f FisheyeProjectionMode::getPickRay(float x, float y, float zoom) const
+{
+    float r = std::hypot(x, y);
+    float phi = celestia::numbers::pi_v<float> * r;
+    float sin_phi = std::sin(phi);
+    float theta = std::atan2(y, x);
+    float newX = sin_phi * std::cos(theta);
+    float newY = sin_phi * std::sin(theta);
+    float newZ = std::cos(phi);
+    Vector3f pickDirection(newX, newY, -newZ);
+    pickDirection.normalize();
+
+    return pickDirection;
+}
+
+bool FisheyeProjectionMode::isFisheye() const
+{
+    return true;
+}
+
+void FisheyeProjectionMode::setScreenDpi(int dpi)
+{
+    // DPI does not affect projection
+}
+
+void FisheyeProjectionMode::setDistanceToScreen(int dts)
+{
+    // Distance to screen does not affect projection
+}
+
+void FisheyeProjectionMode::setSize(float w, float h)
+{
+    width = w;
+    height = h;
+}
+
+std::shared_ptr<ProjectionMode> FisheyeProjectionMode::clone()
+{
+    return std::make_shared<FisheyeProjectionMode>(*this);
```

Comment:
i'm unsure if this works as expected

Reply:
and i don't think that we actually need this.

you have an object that owns ProjectionMode, it should have std::unique_ptr to it. other classes that doesn't own it should use raw pointer.

Reply:
yeah this should be unique_ptr, this is used in setFOVFromZoom/setZoomFromFOV where i do not wish to change renderer's projectionmode's width/height